File: kraken-engine/src/main/java/kraken/runtime/engine/context/data/ExtractedChildDataContextBuilder.java
/*
 *  Copyright 2019 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package kraken.runtime.engine.context.data;

import kraken.runtime.EvaluationSession;
import kraken.runtime.engine.context.StaticContextDataProvider;
import kraken.runtime.expressions.KrakenExpressionEvaluationException;
import kraken.runtime.model.context.ContextNavigation;
import kraken.runtime.engine.context.extraction.ContextChildExtractionInfo;
import kraken.runtime.engine.context.extraction.instance.ContextExtractionResultBuilder;
import kraken.runtime.expressions.KrakenExpressionEvaluator;

import java.util.*;
import java.util.stream.Collectors;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * @author psurinin
 */
public class ExtractedChildDataContextBuilder {

    private final static Logger logger = LoggerFactory.getLogger(ExtractedChildDataContextBuilder.class);

    private final DataContextBuilder dataContextBuilder;

    private final ContextExtractionResultBuilder childBuilder;

    private final KrakenExpressionEvaluator evaluator;

    private final EvaluationSession session;

    public ExtractedChildDataContextBuilder(
            DataContextBuilder dataContextBuilder,
            ContextExtractionResultBuilder childBuilder,
            KrakenExpressionEvaluator evaluator,
            EvaluationSession session
    ) {
        this.dataContextBuilder = dataContextBuilder;
        this.childBuilder = childBuilder;
        this.evaluator = evaluator;
        this.session = session;
    }

    public List<DataContext> resolveImmediateChildren(ContextChildExtractionInfo info) {
        return Optional
                .ofNullable(resolveInheritedChildren(info))
                .orElseGet(() -> resolveExtractedChildren(info));
    }

    private List<DataContext> resolveExtractedChildren(ContextChildExtractionInfo info) {
        ContextNavigation contextNavigation = info.getFrom().getChildren().get(info.getChildContextName());

        Object extractedChild = evaluateNavigationExpression(contextNavigation, info);

        return childBuilder.buildFrom(extractedChild).stream()
            .map(instance -> dataContextBuilder.buildFromExtractedObject(instance, info))
            .collect(Collectors.toList());
    }

    private Object evaluateNavigationExpression(ContextNavigation contextNavigation, ContextChildExtractionInfo info) {
        try {
            return evaluator.evaluateNavigationExpression(contextNavigation, info.getParentDataContext(), session);
        } catch (KrakenExpressionEvaluationException e) {
            logger.trace(
                "Error while evaluating navigation expression: {}. No object will be extracted.",
                contextNavigation.getNavigationExpression().getOriginalExpressionString(),
                e
            );
            return null;
        }
    }

    private List<DataContext> resolveInheritedChildren(ContextChildExtractionInfo info) {
        // if its a step up on inheritance tree, just create data context with same object and different name
        final Collection<String> inheritedContexts = info.getFrom().getInheritedContexts();
        final boolean hasParents = Objects.nonNull(inheritedContexts) && !inheritedContexts.isEmpty();
        // if any of inherited contexts equals next in chain, create copy with its name
        final boolean parentsContainsChild =
                hasParents && inheritedContexts.contains(info.getChildContextName());
        if (hasParents && parentsContainsChild) {
            final DataContext dataContext = dataContextBuilder.buildForAncestorObject(
                    info.getParentDataContext().getDataObject(),
                    info.getChildContextName(),
                    info.getParentDataContext());
            return Collections.singletonList(dataContext);
        }
        return null;
    }
}

File: javascript/packages/engine/src/engine/contexts/data/ExtractedChildDataContextBuilder.ts
/*
 *  Copyright 2018 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
import { toBe } from 'declarative-js'
import { Contexts } from 'kraken-model'
import ContextDefinition = Contexts.ContextDefinition
import { DataContextBuilder } from './DataContextBuilder'
import { requireDefinedValue } from '../../../utils/Utils'
import { ExpressionEvaluator } from '../../runtime/expressions/ExpressionEvaluator'
import { ExpressionEvaluationResult } from 'kraken-engine-api'
import { DataContext } from './DataContext'

export interface ContextChildExtractionInfo {
    parentContextDefinition: ContextDefinition
    parentDataContext: DataContext
    childContextName: string
}

export class ExtractedChildDataContextBuilder {
    constructor(
        private readonly dataContextBuilder: DataContextBuilder,
        private readonly expressionEvaluator: ExpressionEvaluator,
    ) {
        this.resolveImmediateChildren = this.resolveImmediateChildren.bind(this)
        this.extract = this.extract.bind(this)
    }

    resolveImmediateChildren(info: ContextChildExtractionInfo): DataContext[] {
        const { childContextName, parentDataContext } = info
        const result = this.extract(info)
        const toDataContext = this.createDataContextFactory(childContextName, parentDataContext)
        if (Array.isArray(result)) {
            return result.filter(toBe.present).map(toDataContext)
        }
        return [toDataContext(result)]
    }

    private createDataContextFactory(
        childContextName: string,
        parentDataContext: DataContext,
    ): (dataObject: object, index?: number) => DataContext {
        const { dataContextBuilder } = this
        return function _createDataContextFactory(dataObject: object, index?: number): DataContext {
            return dataContextBuilder.buildFromExtractedObject(dataObject, childContextName, parentDataContext, index)
        }
    }
    private extract(info: ContextChildExtractionInfo): object | object[] {
        const children = requireDefinedValue(info.parentContextDefinition.children, 'Children cannot be null')
        const navigation = children[info.childContextName]
        const expressionResult = this.expressionEvaluator.evaluateNavigationExpression(
            navigation,
            info.parentDataContext,
        )

        const extracted = ExpressionEvaluationResult.isError(expressionResult) ? undefined : expressionResult.success

        if (extracted === undefined) {
            return []
        }
        if (Array.isArray(extracted)) {
            return flat(extracted)
        }
        return extracted as object | object[]
    }
}

function flat(array: unknown[]): unknown[] {
    return array.reduce((pv: unknown[], cv) => {
        if (Array.isArray(cv)) {
            return pv.concat(flat(cv))
        }
        if (cv) {
            return pv.concat(cv)
        }
        return pv
    }, [])
}
